runtime/src/enclave_rpc/demux.rs

```diff
@@ -57,164 +48,300 @@ impl From<Error> for crate::types::Error {
     }
 }
 
-/// A map of session identifiers to session instances.
-type SessionMap = HashMap<SessionID, Arc<tokio::sync::Mutex<MultiplexedSession>>>;
+/// Peer identifier.
+type PeerID = Vec<u8>;
+
+/// Shared pointer to a multiplexed session.
+type SharedSession = Arc<tokio::sync::Mutex<MultiplexedSession>>;
+
+/// Key for use in the by-idle-time index.
+type SessionByTimeKey = (i64, PeerID, SessionID);
+
+/// Structure used for session accounting.
+struct SessionMeta {
+    /// Peer identifier.
+    peer_id: PeerID,
+    /// Session identifier.
+    session_id: SessionID,
+    /// Timestamp when the session was last accessed.
+    last_access_time: i64,
+    /// The shared session pointer that needs to be locked for access.
+    inner: SharedSession,
+}
+
+impl SessionMeta {
+    /// Key for ordering in the by-idle-time index.
+    fn by_time_key(&self) -> SessionByTimeKey {
+        (self.last_access_time, self.peer_id.clone(), self.session_id)
+    }
+}
+
+/// Session indices and management operations.
+struct Sessions {
+    /// Session builder.
+    builder: Builder,
+    /// Maximum number of sessions.
+    max_sessions: usize,
+    /// Maximum number of sessions per peer.
+    max_sessions_per_peer: usize,
+    /// Stale session timeout (in seconds).
+    stale_session_timeout: i64,
+
+    /// A map of sessions for each peer.
+    by_peer: HashMap<PeerID, HashMap<SessionID, SessionMeta>>,
+    /// A set of all sessions, ordered by idle time.
+    by_idle_time: BTreeSet<SessionByTimeKey>,
+}
+
+impl Sessions {
+    /// Create a new session management instance.
+    fn new(
+        builder: Builder,
+        max_sessions: usize,
+        max_sessions_per_peer: usize,
+        stale_session_timeout: i64,
+    ) -> Self {
+        Self {
+            builder,
+            max_sessions,
+            max_sessions_per_peer,
+            stale_session_timeout,
+            by_peer: HashMap::new(),
+            by_idle_time: BTreeSet::new(),
+        }
+    }
+
+    /// Create a new multiplexed session.
+    fn create_session(
+        mut builder: Builder,
+        peer_id: PeerID,
+        session_id: SessionID,
+        now: i64,
+    ) -> SessionMeta {
+        // If no quote policy is set, use the local one.
+        if builder.get_quote_policy().is_none() {
```

Comment:
Previous version did not have this `if`. We always want to replace the policy, so that it is up to date. In this case, however, I think that the policy should be always `none`, so it's the same.